### ISSUE TITLE
Cache control for UI assets

### ISSUE THREAD

**Issue by umputun** [umputun/remark42#556]
UI updates may not be picked up by browsers or even worse - picked up partially due to the caching. 

Currently, we let browsers decide what and how to cache and not enforcing any control. I have tried to do it with https://github.com/umputun/remark/commit/b0898cd1f3cd29d3c4fb8ddcc4df370bbb7278d7 and technically it does what I've planned - sets etag as a mix of url & version and checks`If-None-Match` tag matches. Probably it is the right thing to do, but it doesn't help with refresh/update issues.

#554 is trying to address it from the frontend side.

**Comment by paskal** [comment]
I've observed following behavior on reloading http://127.0.0.1:8080/web/remark.js without cache (Cmd + Shift + R on MacOS) using Firefox with added debug logging:

    remark42-dev | 2020/01/18 06:35:14.217 [DEBUG] {api/rest.go:583 api.cacheControl.func2.1} cacheControl request for /web/remark.js
    remark42-dev | 2020/01/18 06:35:14.217 [DEBUG] {api/rest.go:475 api.addFileServer.func1} addFileServer request for /web/remark.js
    remark42-dev | 2020/01/18 06:35:14.217 [DEBUG] {api/rest.go:583 api.cacheControl.func2.1} cacheControl request for /web/remark.js
    remark42-dev | 2020/01/18 06:35:14.217 [DEBUG] {api/rest.go:475 api.addFileServer.func1} addFileServer request for /web/remark.js

For some reason request for a static asset is handled by all involved handlers twice. I've checked the content of the request, and it's the same for both requests.

On reload `addFileServer` request handler is not called because cacheControl return 304 answer, but it's still getting called twice for some reason:

    remark42-dev | 2020/01/18 06:36:07.041 [DEBUG] {api/rest.go:583 api.cacheControl.func2.1} cacheControl request for /web/remark.js
    remark42-dev | 2020/01/18 06:36:07.041 [DEBUG] {api/rest.go:583 api.cacheControl.func2.1} cacheControl request for /web/remark.js

---

However, after a few application rebuilds, I see single request to every endpoint again and couldn't reproduce it. Continuing investigation.

**Comment by paskal** [comment]
Three distinct requests to http:`//127.0.0.1:8080/web/` with Firefox.

### First load after force-reload

| URL                       | `addFileServer` handler call | `cacheControl` call      |
| ------------------------- | -------------------------- | ---------------------- |
| `/web/`                   | ✓                          | no If-None-Match check |
| `/web/counter.js`         | ✓                          | no If-None-Match check |
| `/web/embed.js`           | ✓                          | no If-None-Match check |
| `/web/last-comments.css`  | ✓                          | no If-None-Match check |
| `/web/last-comments.html` | ✓                          | no If-None-Match check |
| `/web/last-comments.js`   | ✓                          | no If-None-Match check |
| `/web/iframe.html?site_id=remark&host=http%3A%2F%2F127.0.0.1%3A8080&url=https%3A%2F%2Fremark42.com%2Fdemo%2F&components=embed%2Ccounter&theme=dark` | ✓ | no If-None-Match check |
| `/web/remark.css`         | ✓                          | no If-None-Match check |
| `/web/remark.js`          | ✓                          | no If-None-Match check |

### Second load after normal reload

| URL                       | `addFileServer` handler call | `cacheControl` call              |
| ------------------------- | -------------------------- | ------------------------------ |
| `/web/`                   | ✓                          | If-None-Match check, cache hit |
| `/web/counter.js`         | ✓                          | If-None-Match check, cache hit |
| `/web/embed.js`           | ✓                          | If-None-Match check, cache hit |
| `/web/last-comments.css`  | ✓                          | If-None-Match check, cache hit |
| `/web/last-comments.html` | ✓                          | If-None-Match check, cache hit |
| `/web/last-comments.js`   | ✓                          | If-None-Match check, cache hit |
| `/web/iframe.html?site_id=remark&host=http%3A%2F%2F127.0.0.1%3A8080&url=https%3A%2F%2Fremark42.com%2Fdemo%2F&components=embed%2Ccounter&theme=dark` | ✓ | no If-None-Match check |

### Third load in a new tab without cleaning cache.

On multiple loads at most I've seen one file from the table below loaded, on most reloads no `addFileServer` request handler is not called.

| URL                      | `addFileServer` handler call | `cacheControl` call      |
| ------------------------ | -------------------------- | ---------------------- |
| `/web/`                  | ✓                          | no If-None-Match check |
| `/web/remark.css`        | ✓                          | no If-None-Match check |
| `/web/last-comments.css` | ✓                          | no If-None-Match check |

## Observation results

`/web/` handling via `addFileServer` looks weird to me, not sure if that's intended.

`/web/iframe.html` handling via `addFileServer` looks weird because it does not seems static, which explains no cache hits.

`/web/remark.css` and `/web/remark.js` are handled differently than `/web/counter.js` and `/web/last-comments.css` and other files mentioned on the second run: they are not getting requested at all from the server when files mentioned in the second load are properly re-requested and get 304 Not Modified response from `cacheControl` middleware.

In Firefox it shows response code 200, cached, but does not ask a server for `/web/remark.css` and `/web/remark.js` at all. I think that is the reason for the strange behavior we observe and aligning these files behavior with other static resources will at least give us consistent behavior which we lack now.

**Comment by paskal** [comment]
What we discovered after a day of tinkering: current cache works as expected in Chrome, after 24h (current Cache-Control max-age value) resources will be re-requested and will get `304 Not Modified` status in case `ETag` head is the same, or `200` in case `ETag` is different.

After setting max-age to a short period like 5 (seconds) Chrome will re-query static resources after this period, but Firefox won't do that because of [iframe cache bug](https://bugzilla.mozilla.org/show_bug.cgi?id=356558) and will reload resources only after page reload (not the force one, normal is enough). With `Cache-Control: must-revalidate` Chrome starts reloading all static resources on every request (and gets `304` for them every time), Firefox doesn't change it's behavior and renders them from cache on load until the user hits reload button.

Our (mine and @akellbl4) suggestion would be to lower max-age to something lower like `1h` and forget about clients who still got cache from pre b0898cd times because there is no way to affect them with whatever we are doing at this very moment: they will catch up with new caching policy eventually.

**Comment by paskal** [comment]
And the conclusion is that the only reliable way of delivering changes as soon as they are ready and not with some delay (taking into account the Firefox bug) is to have remark version embedded somewhere in the path of the all static resources (as having just different parameters for iframe.html won't work in Firefox).

**Comment by umputun** [comment]
> For some reason request for a static asset is handled by all involved handlers twice. I've checked the content of the request, and it's the same for both requests.

I have tried to reproduce this and could now see the issue. One request handled once in all my tests. However, I wonder where exactly you put logging? What I did was just adding this line:

<img width="1076" alt="3ri7khmxj9-20200119-003252" src="https://user-images.githubusercontent.com/535880/72676114-4276cd00-3a53-11ea-943d-abc56ba78f7d.png">

I have tested http://127.0.0.1:8080/web/remark.js calls from safari and got no 304, each one was 200. From chrome, it worked as expected (200 first, 304 after this). I have checked what other cache-related headers sites send, and the only one I found is `Expires`. Tried to add `w.Header().Set("Expires", time.Now().Add(expiration).UTC().Format(time.RFC1123))`
 and it changed nothing I can see.

**Comment by paskal** [comment]
> I have tried to reproduce this and could not see the issue. One request handled once in all my tests. However, I wonder where exactly you put logging?

This is the debug logging I've used: paskal/remark/pull/7.

> I have checked what other cache-related headers sites send, and the only one I found is `Expires`.
 
I was testing in Firefox and Chrome and now checked Safari: I don't see `Expires` tag added anywhere in our static files Go serving.

Also, it's expected to cache file for max-age period and don't request it from the server. This is overwritten by Firefox reload (which reloads everything on the page) and Chrome reloads (which reloads only current HTML page, and maybe resources on it if page changed itself - don't know yet). Safari seems to be more aggressive and doesn't reload cached resources from server on the reload button hit.

**Comment by paskal** [comment]
The only summary I could come up with: if we want consistently deliver the latest version of the UI assets to the user without any delay, we should generate unique file names for all assets likely by embedding commit hash into them.

After research, I've done previously I don't have any ideas about what else could be done here.

Facebook does that, and even Google does that on their main page that despite having control over the most popular browser in the world. I think despite our opinion on how ugly this hack is, we should either give up on ~1h+ delay in assets refresh, on go on with unique filenames path and solve it that way.

**Comment by umputun** [comment]
I think we fine as-is. I.e. "give up on ~1h+ delay in assets refresh" seems to be a reasonable tradeoff

**Comment by akellbl4** [comment]
@umputun our last changes on cache should fix this issue.

**Comment by umputun** [comment]
agree

done in 0dd3cd93ee6702a6351691a573f9c8dfab76a3e4 12b0fbe75d018911e33669080a15b5886f864e8b and 6d525e48208d4526147928e1d27e7d7b3804500a
